Source/Diagnostics/ReducedDiags/ParticleHistogram.cpp

```diff
@@ -159,30 +159,31 @@ void ParticleHistogram::ComputeDiags (int step)
     const bool is_unity_particle_weight =
         (m_norm == NormalizationType::unity_particle_weight) ? true : false;
 
-    for ( int i = 0; i < m_bin_num; ++i )
-    {
-        // compute the histogram
-        m_data[i] = ReduceSum( myspc,
-        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
-        {
-            // return 0 if the particle is filtered out
-            if (!fun_filterparser(p, amrex::RandomEngine())) {return 0.0_rt;}
-            // continue function if particle is not filtered out
-            auto const w  = p.rdata(PIdx::w);
-            amrex::ParticleReal x, y, z;
-            get_particle_position(p, x, y, z);
-            auto const ux = p.rdata(PIdx::ux)/PhysConst::c;
-            auto const uy = p.rdata(PIdx::uy)/PhysConst::c;
-            auto const uz = p.rdata(PIdx::uz)/PhysConst::c;
-            auto const f = fun_partparser(t,x,y,z,ux,uy,uz);
-            auto const f1 = bin_min + bin_size*i;
-            auto const f2 = bin_min + bin_size*(i+1);
-            if ( f > f1 && f < f2 ) {
-                if ( is_unity_particle_weight ) return 1.0_rt;
-                else return w;
-            } else return 0.0_rt;
-        });
+    // compute the histogram
+    for ( int i = 0; i < m_bin_num; ++i ) {
+        m_data[i] = 0.0_rt;
     }
+    ReduceSum( myspc,
+    [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+    {
+        // return 0 if the particle is filtered out
+        if (!fun_filterparser(p, amrex::RandomEngine())) {return 0.0_rt;}
+        // continue function if particle is not filtered out
+        auto const w  = p.rdata(PIdx::w);
+        amrex::ParticleReal x, y, z;
+        get_particle_position(p, x, y, z);
+        auto const ux = p.rdata(PIdx::ux)/PhysConst::c;
+        auto const uy = p.rdata(PIdx::uy)/PhysConst::c;
+        auto const uz = p.rdata(PIdx::uz)/PhysConst::c;
+        auto const f = fun_partparser(t,x,y,z,ux,uy,uz);
+        int i = int(std::floor((f-bin_min)/bin_size));
```

Comment:
Avoiding memory violations if particles are outside of the set min/max range:
```suggestion
        int i = int(std::floor((f-bin_min)/bin_size));

        // all entries larger than bin_max go into the last bin */
        i = i < m_bin_num ? i : m_bin_num-1;
        // all entries smaller than bin_min go into the first bin
        i = i > 0 ? i : 0;
```

Reply:
@ax3l 
I actually think it is more reasonable to ignore those particles that are outside of the set.
Previously, I often intent to set a smaller range, so it actually behaves like a filter to some extent.

Reply:
Both is possible, but even in that case you need to handle it by not calling `atomicAdd` anymore.

But since we have actual filters now, I would change the behavior - overflow and underflow bins are extremely useful.
We could add +2 bins for those as many histogram libs do as well (one at the beginning and one at the end of the spectrum).

Reply:
Update: `if ( i<0 || i>=m_bin_num )` handling is ok for now, we can add overflow bins in a follow-up.